cryptoadvance/specter-desktop
# Add "Abandon transaction" option for low fee txs that have been purged from the mempool

Fixes #951

<img width="973" alt="abandon_tx_screenshot" src="https://user-images.githubusercontent.com/934746/109696259-77a88680-7b52-11eb-9013-020f15764e0f.png">

Changes/Rationale:
`DEVELOPMENT.md`:
* Additional notes for running bitcoind manually.
* minor reorg.

`bitcoind.py`:
* Now takes optional `extra_args` when starting bitcoind to allow for setting `maxmempool` or other future test case needs.
* Implement stop_bitcoind for BitcoindPlainController. Running the tests against local bitcoind was leaving the process running.

`bitcoin_core.py`:
* Adds `sign_raw_tx` helper for the test case. Parallels `sign_psbt`.

`wallets.py`:
* Adds `is_purged` check when loading the tx details popup.
* 'Abandon transaction' button posts to `history` endpoint to keep the flow simple (no Ajax back to the tx details popup).

`tx-data.html`:
* The "Abandon transaction" text tries to strike a balance between being informative for less savvy users and guide them on when to properly use the feature AND be as brief as possible! I did my best!

~~`wallet-history.jinja`:~~
~~* I needed the `csrf_token` so I can do the POST on the "Abandon transaction" click.~~

`wallet.py`:
* The additional `if "address" in tx` check was necessary to get through the test case. I think the bitcoin core method of generating large transactions makes them nonstandard (also why we pass in `acceptnonstdtxn=1` when starting bitcoind)?

`test/bitcoin_core/*`
* See the README added there.

`conftest.py`:
* Factored `instantiate_bitcoind_controller` out of `bitcoin_regtest` so it could be re-used with different `extra_args` as needed for this test and potentially future tests.
* Pass `rpcport` into `BitcoindPlainController` so it can differentiate between the default port and the one-off instance used in the test.

`test_wallet_manager`:
* First two tests were failing against a local bitcoind but passing with the docker version. Changed the tests to use the new `instantiate_bitcoind_controller` so they can run on a completely fresh, isolated bitcoind.


HOW TO TEST ON MAINNET:
Restart your node. The mempool should reinitialize `mempoolminfee` to 0.00001000 (check via `bitcoin-cli getmempoolinfo`). Create a transaction with 1sat/vB fee and broadcast it. If there are enough txs already in other nodes' mempools, odds are that they will ignore your tiny transaction for violating their `mempoolminfee`. Hit a public block explorer to confirm that your txid never landed. But the tx will still be in your node's mempool.

Restart your node again but this time pass in `-maxmempool=5` to severely restrict the size of your mempool. Watch the `mempoolminfee` creep up as the node fills its limited mempool with just the highest-fee txs. Anything above 0.00001000 should prompt your node to purge your tiny tx. Specter should now show you the "Abandon transaction" option.

Click to abandon the tx to clear it out of your wallet. Restart your node with normal settings. Send a new test tx using the same coins but with a reasonable fee for the current mempool. The abandoned tx should no longer appear in your history, just the new tx. This final tx should land a confirmation eventually and get updated in Specter accordingly.


FOLLOW-UP WORK TODO:
The abandoned tx still shows up on the history page. A different icon or status indicator would be useful in the list as well as in the `decoderawtx` details popup. The abandoned tx disappears once its coins are spent in a new tx.
